src/debugger/repl.jl

```diff
@@ -0,0 +1,94 @@
+using REPL
+using REPL.LineEdit
+using REPL.REPLCompletions
+using JuliaInterpreter: moduleof, locals, eval_code
+
+const normal_prefix = Sys.iswindows() ? "\e[33m" : "\e[38;5;166m"
+const compiled_prefix = "\e[96m"
+
+function debugprompt()
+  try
+    panel = REPL.LineEdit.Prompt("debug> ";
+              prompt_prefix = isCompileMode() ? compiled_prefix : normal_prefix,
+              prompt_suffix = Base.text_colors[:normal],
+              complete = JunoDebuggerRPELCompletionProvider(),
+              on_enter = s -> true)
+
+    panel.hist = REPL.REPLHistoryProvider(Dict{Symbol,Any}(:junodebug => panel))
+    REPL.history_reset_state(panel.hist)
+
+    search_prompt, skeymap = LineEdit.setup_search_keymap(panel.hist)
+    search_prompt.complete = REPL.LatexCompletions()
+
+    panel.on_done = (s, buf, ok) -> begin
+      if !ok
+        LineEdit.transition(s, :abort)
+        REPL.LineEdit.reset_state(s)
+        return false
+      end
+      Atom.msg("working")
+
+      line = String(take!(buf))
+
+      if isempty(line)
+        Atom.msg("doneWorking")
+        return true
+      end
+
+      try
+        r = interpret(line)
+        r ≠ nothing && display(r)
+      catch err
+        display_error(stderr, err, stacktrace(catch_backtrace()))
+      end
+      println()
+      LineEdit.reset_state(s)
+
+      Atom.msg("doneWorking")
+      Atom.msg("updateWorkspace")
+
+      return true
+    end
+
+    panel.keymap_dict = LineEdit.keymap(Dict{Any,Any}[skeymap, LineEdit.history_keymap, LineEdit.default_keymap, LineEdit.escape_defaults])
+
+    REPL.run_interface(Base.active_repl.t, REPL.LineEdit.ModalInterface([panel, search_prompt]))
+  catch e
+    Atom.msg("doneWorking")
+    Atom.msg("updateWorkspace")
+    e isa InterruptException || rethrow(e)
+  end
+end
+
+# completions
+
+struct JunoDebuggerRPELCompletionProvider <: REPL.CompletionProvider end
+
+function LineEdit.complete_line(c::JunoDebuggerRPELCompletionProvider, s)
+  partial = REPL.beforecursor(s.input_buffer)
+  full = LineEdit.input_string(s)
+
+  frame = STATE.frame
+
+  # module-aware repl backend completions
+  comps, range, should_complete = REPLCompletions.completions(full, lastindex(partial), moduleof(frame))
+  ret = map(REPLCompletions.completion_text, comps)
+
+  # local completions -- should be shown first
+  @>> locals(frame) filter!(v -> begin
+    # ref: https://github.com/JuliaDebug/JuliaInterpreter.jl/blob/master/src/utils.jl#L365-L370
+    if v.name == Symbol("#self") && (v.value isa Type || sizeof(v.value) == 0)
+      return false
+    else
+      return startswith(string(v.name), partial)
+    end
+  end) map(v -> string(v.name)) prepend!(ret) unique!
+
+  return ret, partial[range], should_complete
+end
+
+# Evaluation
+function interpret(code::AbstractString, state::DebuggerState = STATE)
+  state.frame === nothing && return
+  eval_code(active_frame(state), code)
```

Comment:
I tried to understand the behaviour of `active_frame` and `STATE.level`, but I'm still not sure why `active_frame` is needed.

Is there a case when `STATE.frame` doesn't point to the innermost (i.e. active) frame ?
Should I need to use this logic in datatip as well ?

Reply:
So `STATE.frame` is the frame we stopped interpreting in, but you can also switch to a parent frame by clicking on the stackframe number (which is saved in `STATE.level`). Code is then evaluated in that context. So yes, the debugging datatips need `active_frame(state)` instead of `state.frame` as well.

Reply:
I see, I missed to check the `setStackLevel` handler.

So then maybe using `active_frame` in
- repl completions
- workspace (only show local binds up to the `active_frame(state)`)
- datatip

would be more consistent I think ?

(I included the change in f188e4c)